### ISSUE TITLE
cc @daxian-dbw to provide feedback regarding PS7

### ISSUE THREAD

**Issue by iRon7** [dotnet/runtime#63537]
cc @daxian-dbw to provide feedback regarding PS7

_Originally posted by @SteveL-MSFT in https://github.com/dotnet/runtime/issues/56835#issuecomment-944409561_

**Comment by iRon7** [comment]
From the questions in the initial issue #56835 [Make OrderedDictionaryKeyValueCollection implement IList](https://github.com/dotnet/runtime/issues/56835), I was under the impression that all the key lists of the dictionaries under the [System.Collections.Generic Namespace]( https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic) and [System.Collections.Specialized Namespace]( https://docs.microsoft.com/en-us/dotnet/api/system.collections.specialized)  would get an `IList` interface but from issue #60060  [issue 56835 make ordered dictionary kv collection implement i list]( https://github.com/dotnet/runtime/pull/60060) it looks like the implementation is actually limited to just the [OrderedDictionary Class]( https://docs.microsoft.com/ dotnet/api/system.collections.specialized.ordereddictionary).

[As stated](https://github.com/dotnet/runtime/issues/56835#issuecomment-937854890) in #56835:


> The `System.Collections.Specialized.OrderedDictionary` is most used in specialized collection in PowerShell (as it is accelerated with `[Ordered]`) but not limited to it.
> So, yes, ***it would be nice to have this on all other specialized collection key lists as well***.
> This would simplify the PowerShell syntax from e.g. `@($Dictionary.Keys)[0]` to `$Dictionary.Keys[0]`.

Besides, it will be more consistent to do this for all concerned dictionaries to prevent any confusion why the following (is going to) works for an `OrderedDictionary`

```PowerShell
$Ordered = [System.Collections.Specialized.OrderedDictionary]::New()
$Ordered.Add('b', 2)
$Ordered.Add('a', 1)
$Ordered.Keys[1]
```
But doesn’t for e.g. a `SortedDictionary`:

```PowerShell
$Sorted = [System.Collections.Generic.SortedDictionary[String,Int]]::New()
$Sorted.Add('b', 2)
$Sorted.Add('a', 1)
$Sorted.Keys[1
```

<sub>(I should be noted that although `[ System.Collections.Specialized.OrderedDictionary]` has a type accelerator in PowerShell, there are reasons to fall back on the full qualified .Net name to e.g. construct a CaseSensitive dictionary)</sub>

Only for unordered dictionaries (as e.g. the [Hashtable Class](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable)) it might makes less sense to implement an `IList` interface as the keys are unordered by definition (but might still done for consistency).

In other words:
Can the `IList` interface be implemented for all (ordered) dictionary keys?

**Comment by iRon7** [comment]
> Only for _unordered_ dictionaries (as e.g. the [Hashtable Class](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable)) it might makes less sense to implement an `IList` interface as the keys are unordered by definition (but might still done for consistency).

Actually, even for *unordered* directory keys there could be an advantage in PowerShell when the `IList` interface is implemented as someone might want to simply use *all* properties keys (rather than a specific one) and not care about the order:

Current syntax:
```PowerShell
$Properties = @{ a = 1; b = 2 }
$Object = [PSCustomObject]@Properties
$Object |Select-Object -Property @($Properties.Keys)
```
Expected syntax:
```PowerShell
$Object |Select-Object -Property $Properties.Keys
```

**Comment by eiriktsarpalis** [comment]
See also https://github.com/dotnet/runtime/pull/60060#issuecomment-936014202.

We would need a champion to do the prep work, enumerating all the collections and proposed changes. @iRon7 would you be willing to take this on?

> Actually, even for unordered directory keys there could be an advantage in PowerShell when the IList interface is implemented as someone might want to simply use all properties keys (rather than a specific one) and not care about the order

I see the point but I don't think it is a good idea to implement IList for collections with nondeterministic ordering.

**Comment by daxian-dbw** [comment]
I think it makes sense for `SortedDictionary`, but for collections with nondeterministic ordering, I agree that it makes less sense as I don't see the usefulness of it.

@iRon7 The example you shared in your comment has a parsing error. But regardless of that, it doesn't seem to be a real scenario given that `$object` would essentially be the same as `$Object |Select-Object -Property @($Properties.Keys)`.

**Comment by iRon7** [comment]
@eiriktsarpalis,
> We would need a champion to do the prep work, enumerating all the collections and proposed changes. @iRon7 would you be willing to take this on?

I have no in-depth C# knowkledge. My knowledge of .Net comes mainly from PowerShell.
Anyways, I have browsed through all the `System.Collection` classes and found the following (key) lists:
From which 5 (`ArrayList`, `List`, `SortedList`, `NameValueCollection`, `StringCollection`) have the `IList` interface implemented:
(@mklement0, can you think of any other collection that might/should support the `IList` interface?)

| Collection                                                      | KeyCollection    | Ordered | IList |
| --------------------------------------------------------------- | ---------------- | ------- | ----- |
| System.Collections.ArrayList                                    | ArrayList        | ✔️       | ✔️    |
| System.Collections.Hashtable<TKey,TValue>                       | Keys             | ❌       | ❌    |
| System.Collections.Generic.HashSet<T>                           | HashSet          | ✔️       | ❌    |
| System.Collections.Generic.LinkedList<T>                        | LinkedList       | ✔️       | ❌    |
| System.Collections.Generic.List<T>                              | List             | ✔️       | ✔️    |
| System.Collections.Generic.Queue<T>                             | Queue            | ✔️       | ❌    |
| System.Collections.Generic.Dictionary<TKey,TValue>              | Keys             | ❌️       | ❌    |
| System.Collections.Generic.SortedDictionary<TKey,TValue>        | Keys             | ✔️       | ❌    |
| System.Collections.Generic.SortedList<TKey,TValue>              | Keys             | ✔️       | ✔️    |
| System.Collections.Generic.SortedSet<T>                         | SortedSet        | ✔️       | ❌    |
| System.Collections.Generic.Stack<T>                             | Stack            | ✔️       | ❌    |
| System.Collections.Specialized.HybridDictionary<TKey,TValue>    | Keys             | ❌️       | ❌    |
| System.Collections.Specialized.ListDictionary<TKey,TValue>      | Keys             | ❌️       | ❌    |
| System.Collections.Specialized.NameValueCollection<TKey,TValue> | Keys             | ❌️       | ✔️    |
| System.Collections.Specialized.OrderedDictionary                | Keys             | ✔️       | ❌    |
| System.Collections.Specialized.StringCollection                 | StringCollection | ✔️       | ✔️    |
| System.Collections.Specialized.StringDictionary                 | Keys             | ✔️       | ❌    |

I have used the following **PowerShell** to confirm the `IList` interface implementation:

<details>
  <summary>PowerShell IList tests</summary>
  
```PowerShell
#Requires -Modules @{ModuleName="Pester"; ModuleVersion="5.0.0"}

# System.Collections Namespace

$ArrayList = [System.Collections.ArrayList]::new() # Ordered
$Null = $ArrayList.Add('a')
$Null = $ArrayList.Add('b')
$ArrayList[1] |Should -Not -BeNullOrEmpty # Ok

# $Hashtable = @{ a = 1; b = 2 }
$Hashtable = [System.Collections.Hashtable]::new() # Unordered
$Hashtable.Add('a', 1)
$Hashtable.Add('b', 2)
$Hashtable.Keys[1] |Should -Not -BeNullOrEmpty # Expected a value

# System.Collections.Generic Namespace

$HashSet = [System.Collections.Generic.HashSet[String]]::new() # Unordered
$HashSet.Add('a')
$HashSet.Add('b')
$HashSet[1] |Should -Not -BeNullOrEmpty # Expected a value

$LinkedList = [System.Collections.Generic.LinkedList[String]]::new() # Unordered
$LinkedList.Add('a')
$LinkedList.Add('b')
$LinkedList[1] |Should -Not -BeNullOrEmpty # Expected a value

$List = [System.Collections.Generic.List[String]]::new() # ordered
$List.Add('a')
$List.Add('b')
$List[1] |Should -Not -BeNullOrEmpty # Ok

$Queue = [System.Collections.Generic.Queue[String]]::new() # ordered
$Queue.Enqueue('a')
$Queue.Enqueue('b')
$Queue[1] |Should -Not -BeNullOrEmpty # Expected a value

$Dictionary = [System.Collections.Generic.Dictionary[String,Int]]::new() # Unordered
$Dictionary.Add('a', 1)
$Dictionary.Add('b', 2)
$Dictionary.Keys[1] |Should -Not -BeNullOrEmpty # Expected a value

$SortedDictionary = [System.Collections.Generic.SortedDictionary[String,Int]]::New() # Ordered (sorted)
$SortedDictionary.Add('a', 1)
$SortedDictionary.Add('b', 2)
$SortedDictionary.Keys[1] |Should -Not -BeNullOrEmpty # Expected a value

$SortedList = [System.Collections.Generic.SortedList[String,Int]]::New() # Ordered (sorted)
$SortedList.Add('a', 1)
$SortedList.Add('b', 2)
$SortedList.Keys[1] |Should -Not -BeNullOrEmpty # Ok

$SortedSet = [System.Collections.Generic.SortedSet[String]]::new() # ordered (sorted)
$Null = $SortedSet.Add('a')
$Null = $SortedSet.Add('b')
$SortedSet[1] |Should -Not -BeNullOrEmpty # Expected a value

$Stack = [System.Collections.Generic.Stack[String]]::new() # ordered (reversed)
$Stack.Push('a')
$Stack.Push('b')
$Stack[1] |Should -Not -BeNullOrEmpty # Expected a value

# System.Collections.Specialized Namespace

$HybridDictionary = [System.Collections.Specialized.HybridDictionary]::New() # Unordered
$HybridDictionary.Add('a', 1)
$HybridDictionary.Add('b', 2)
$HybridDictionary.Keys[1] |Should -Not -BeNullOrEmpty # Expected a value

$ListDictionary = [System.Collections.Specialized.ListDictionary]::New() # Unordered
$ListDictionary.Add('a', 1)
$ListDictionary.Add('b', 2)
$ListDictionary.Keys[1] |Should -Not -BeNullOrEmpty # Expected a value

$NameValueCollection = [System.Collections.Specialized.NameValueCollection]::New() # Unordered
$NameValueCollection.Add('a', 1)
$NameValueCollection.Add('b', 2)
$NameValueCollection.Keys[1] |Should -Not -BeNullOrEmpty # Ok

# $Ordered = [Ordered]@{ a = 1; b = 2 }
$Ordered = [System.Collections.Specialized.OrderedDictionary]::New() # Ordered
$Ordered.Add('a', 1)
$Ordered.Add('b', 2)
$Ordered.Keys[1] |Should -Not -BeNullOrEmpty # Expected a value

$StringCollection = [System.Collections.Specialized.StringCollection]::New() # Ordered
$Null = $StringCollection.Add('a')
$Null = $StringCollection.Add('b')
$StringCollection[1] |Should -Not -BeNullOrEmpty # Ok

$StringDictionary = [System.Collections.Specialized.StringDictionary]::New() # Ordered
$StringDictionary.Add('a', 1)
$StringDictionary.Add('b', 2)
$StringDictionary.Keys[1] |Should -Not -BeNullOrEmpty # Expected a value
```
</details>

@daxian-dbw 
> The example you shared in your comment has a parsing error.

Thanks, I have fixed it in the concerned comment.

> it doesn't seem to be a real scenario...

You are right, it is just to show how a complete `KeyCollection` might be used. A more real scenario would be if a the `KeyCollection` of another hashtable (with a different key set) would be used.

A use case in PowerShell is where you want to [unite the properties](https://github.com/PowerShell/PowerShell/issues/13906) based on a key collection that is known to contain all the required keys:

```PowerShell
$Properties1 = @{ a = 1; b = 2 }
$Properties2 = @{ a = 3; b = 4 ; c = 5}
[PSCustomObject]$Properties1, [PSCustomObject]$Properties2 |Select-Object @($Properties2.Keys)
```

Wishful thinking:
```PowerShell
[PSCustomObject]$Properties1, [PSCustomObject]$Properties2 |Select-Object $Properties2.Keys
```
